commcare_connect/opportunity/views.py

```diff
@@ -1376,3 +1363,284 @@ def sync_deliver_units(request, org_slug, opp_id):
+            args=[request.org.slug, self.kwargs["opp_id"], self.kwargs["pk"]],
+        )
+        query_params = request.GET.urlencode()
+        response["HX-Replace-Url"] = f"{url}?{query_params}"
+        response["HX-Trigger"] = json.dumps({"load-total-pages": self.table.paginator.num_pages})
+        return response
+
+    def get_table_kwargs(self):
+        kwargs = super().get_table_kwargs()
+        kwargs["organization"] = self.request.org
+        return kwargs
+
+    def get_queryset(self):
+        self.opportunity = get_opportunity_or_404(self.kwargs["opp_id"], self.kwargs["org_slug"])
+        self.is_program_manager = self.request.org.program_manager and (
+            (self.request.org_membership != None and self.request.org_membership.is_admin)  # noqa: E711
+            or self.request.user.is_superuser
+        )
+
+        self.filter_status = self.request.GET.get("filter_status")
+        self.filter_date = self.request.GET.get("filter_date")
+        filter_kwargs = {"opportunity_access": self.kwargs["pk"]}
+        if self.filter_date:
+            date = datetime.datetime.strptime(self.filter_date, "%Y-%m-%d")
+            filter_kwargs.update({"visit_date__date": date})
+
+        if self.filter_status == "pending":
+            filter_kwargs.update({"status": VisitValidationStatus.pending})
+            self.exclude_columns = ["last_activity"]
+        if self.filter_status == "approved":
+            filter_kwargs.update(
+                {
+                    "review_status": VisitReviewStatus.agree,
+                    "status": VisitValidationStatus.approved,
+                }
+            )
+        if self.filter_status == "rejected":
+            filter_kwargs.update({"status": VisitValidationStatus.rejected})
+
+        if self.filter_status == "pending_review":
+            filter_kwargs.update(
+                {
+                    "review_status": VisitReviewStatus.pending,
+                    "status": VisitValidationStatus.approved,
+                }
+            )
+        if self.filter_status == "disagree":
+            filter_kwargs.update(
+                {
+                    "review_status": VisitReviewStatus.disagree,
+                    "status": VisitValidationStatus.approved,
+                }
+            )
+        if self.filter_status == "agree":
+            filter_kwargs.update(
+                {
+                    "review_status": VisitReviewStatus.agree,
+                    "status": VisitValidationStatus.approved,
+                    "flagged": True,
+                }
+            )
+
+        if self.is_program_manager:
+            filter_kwargs.update({"review_created_on__isnull": False})
+        return UserVisit.objects.filter(**filter_kwargs).order_by("visit_date")
+
+
+@org_member_required
+def user_visit_details(request, org_slug, opp_id, pk):
+    opportunity = get_opportunity_or_404(opp_id, org_slug)
+    is_program_manager = request.org.program_manager and (
```

Comment:
this appears to be replicated in a lot of places, would be good to break into shared functionality.

Reply:
Fixed. https://github.com/dimagi/commcare-connect/commit/1fd8f0b01a680543e6060b35fdaaa508928f61ca